File: consistent_hashing.py
import hashlib
import bisect
import struct
from collections import defaultdict

# --- 1. HashFunction Module ---
class HashFunction:
    """
    Handles deterministic hashing using SHA-1.
    """
    def hash(self, key: str) -> int:
        """
        Returns a large integer hash for a given key string.
        Using SHA-1 and taking the first 8 bytes implies a 64-bit keyspace.
        """
        # Encode string to bytes, hash using SHA-1
        hash_bytes = hashlib.sha1(key.encode('utf-8')).digest()
        # Unpack first 8 bytes as a big-endian unsigned long long (0 -> 2^64 - 1)
        return struct.unpack('>Q', hash_bytes[:8])[0]

# --- 2. Node & HashRing Module ---
class ConsistentHashRing:
    """
    Manages the Hash Ring, Node placement, and Lookups.
    """
    def __init__(self, replicas=100):
        """
        replicas: Number of virtual nodes (vnodes) per physical node.
                  Higher K = better load distribution.
        """
        self.replicas = replicas
        self.hash_func = HashFunction()
        self.ring = {}           # Map: hash_val -> physical_node_name
        self.sorted_keys = []    # Sorted list of hash values for O(log N) search
        self.nodes = set()       # Track physical nodes

    def add_node(self, node: str):
        """
        Adds a physical node and its virtual nodes to the ring.
        """
        self.nodes.add(node)
        for i in range(self.replicas):
            # Create a virtual node key, e.g., "NodeA#0", "NodeA#1"
            vnode_key = f"{node}#{i}"
            hash_val = self.hash_func.hash(vnode_key)
            self.ring[hash_val] = node
            # Insert into sorted list while maintaining order
            bisect.insort(self.sorted_keys, hash_val)

    def remove_node(self, node: str):
        """
        Removes a physical node and all its virtual nodes.
        """
        if node not in self.nodes:
            return
        
        self.nodes.remove(node)
        for i in range(self.replicas):
            vnode_key = f"{node}#{i}"
            hash_val = self.hash_func.hash(vnode_key)
            # Remove from ring map
            if hash_val in self.ring:
                del self.ring[hash_val]
            # Remove from sorted keys (O(N) operation in list, acceptable for infrequent config changes)
            # Note: In production, a balanced BST or red-black tree is preferred for faster removal.
            idx = bisect.bisect_left(self.sorted_keys, hash_val)
            if idx < len(self.sorted_keys) and self.sorted_keys[idx] == hash_val:
                self.sorted_keys.pop(idx)

    def get_node(self, key: str) -> str:
        """
        Given a key, find the responsible physical node.
        Time Complexity: O(log N) where N is total vnodes.
        """
        if not self.ring:
            return None
        
        hash_val = self.hash_func.hash(key)
        
        # Binary search for the first hash on the ring >= key's hash
        idx = bisect.bisect_left(self.sorted_keys, hash_val)
        
        # If we reach the end of the list, wrap around to index 0 (Circular Ring)
        if idx == len(self.sorted_keys):
            idx = 0
            
        return self.ring[self.sorted_keys[idx]]

# --- 3. Rebalancer / Evaluator Module ---
class LoadBalancerSimulator:
    """
    Simulates key distribution and measures rebalancing efficiency.
    """
    def __init__(self, num_keys=100000):
        self.num_keys = num_keys
        # Generate stable mock keys (Key0, Key1, ...)
        self.keys = [f"Key{i}" for i in range(num_keys)]

    def get_distribution(self, ring):
        """
        Calculates how many keys are mapped to each node.
        """
        distribution = defaultdict(int)
        for key in self.keys:
            node = ring.get_node(key)
            distribution[node] += 1
        return distribution

    def calculate_diff(self, dist_before, dist_after):
        """
        Calculates the percentage of keys that moved to a DIFFERENT node.
        """
        # To do this accurately, we must re-map every key and check mapping
        # We cannot just compare counts.
        pass # Helper only, actual logic in run_simulation

    def run_scenario(self):
        print("--- Consistent Hashing Simulation ---")
        
        # 1. Setup Initial Ring
        ring = ConsistentHashRing(replicas=100) # K=100 vnodes
        initial_nodes = ["Node_A", "Node_B", "Node_C"]
        for n in initial_nodes:
            ring.add_node(n)
        
        print(f"1. Initial State: {len(initial_nodes)} Nodes, {ring.replicas} vnodes/node")
        
        # 2. Measure Initial Load
        mapping_before = {}
        dist_before = defaultdict(int)
        for key in self.keys:
            node = ring.get_node(key)
            mapping_before[key] = node
            dist_before[node] += 1
            
        # Print Stats
        print(f"   Total Keys: {self.num_keys}")
        print("   Load Distribution:")
        for node, count in dist_before.items():
            print(f"     {node}: {count} keys ({count/self.num_keys:.2%})")

        # 3. Add a Node (Node_D)
        print("\n2. Adding Node_D...")
        ring.add_node("Node_D")
        
        moved_keys = 0
        dist_after_add = defaultdict(int)
        for key in self.keys:
            new_node = ring.get_node(key)
            dist_after_add[new_node] += 1
            if new_node != mapping_before[key]:
                moved_keys += 1
                
        print("   Load Distribution After Add:")
        for node, count in dist_after_add.items():
            print(f"     {node}: {count} keys ({count/self.num_keys:.2%})")
            
        percent_moved = moved_keys / self.num_keys
        ideal_movement = 1.0 / (len(initial_nodes) + 1) # 1/(N+1)
        print(f"   Keys Moved: {moved_keys} ({percent_moved:.2%})")
        print(f"   Theoretical Ideal Movement: {ideal_movement:.2%}")

        # 4. Remove a Node (Node_A)
        print("\n3. Removing Node_A...")
        # Update mapping baseline to current state (with D included)
        mapping_before_remove = {}
        for key in self.keys:
            mapping_before_remove[key] = ring.get_node(key)
            
        ring.remove_node("Node_A")
        
        moved_keys_rem = 0
        dist_after_rem = defaultdict(int)
        for key in self.keys:
            new_node = ring.get_node(key)
            dist_after_rem[new_node] += 1
            if new_node != mapping_before_remove[key]:
                moved_keys_rem += 1

        print("   Load Distribution After Removal:")
        for node, count in dist_after_rem.items():
            print(f"     {node}: {count} keys ({count/self.num_keys:.2%})")
            
        print(f"   Keys Moved: {moved_keys_rem} ({moved_keys_rem/self.num_keys:.2%})")

# --- 4. Main Execution ---
if __name__ == "__main__":
    sim = LoadBalancerSimulator(num_keys=100_000)
    sim.run_scenario()

File: ConsistentHashing.java
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

// --- 1. HashFunction Module ---
class HashFunction {
    private final MessageDigest md;

    public HashFunction() {
        try {
            // Using SHA-1 for better distribution than MD5, deterministic output
            this.md = MessageDigest.getInstance("SHA-1");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-1 algorithm not found", e);
        }
    }

    /**
     * Returns a 64-bit deterministic hash for a given key.
     * Maps the key to the ring space: Long.MIN_VALUE to Long.MAX_VALUE
     */
    public long hash(String key) {
        md.reset();
        md.update(key.getBytes());
        byte[] digest = md.digest();
        // Use the first 8 bytes to create a long
        return ByteBuffer.wrap(digest).getLong();
    }
}

// --- 2. Node Module ---
class Node {
    private final String name;
    
    public Node(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    /**
     * Generates virtual node keys (e.g., "NodeA#0", "NodeA#1")
     */
    public List<String> getVNodeKeys(int replicas) {
        List<String> keys = new ArrayList<>();
        for (int i = 0; i < replicas; i++) {
            keys.add(name + "#" + i);
        }
        return keys;
    }

    @Override
    public String toString() {
        return name;
    }
}

// --- 3. HashRing Module ---
class ConsistentHashRing {
    private final int replicas;
    private final HashFunction hashFunction;
    // TreeMap provides O(log N) lookup and automatic sorting
    private final TreeMap<Long, Node> ring; 

    public ConsistentHashRing(int replicas) {
        this.replicas = replicas;
        this.hashFunction = new HashFunction();
        this.ring = new TreeMap<>();
    }

    public void addNode(String nodeName) {
        Node node = new Node(nodeName);
        for (String vNodeKey : node.getVNodeKeys(replicas)) {
            long hash = hashFunction.hash(vNodeKey);
            ring.put(hash, node);
        }
    }

    public void removeNode(String nodeName) {
        // Iterator avoids ConcurrentModificationException during removal
        Iterator<Map.Entry<Long, Node>> it = ring.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<Long, Node> entry = it.next();
            if (entry.getValue().getName().equals(nodeName)) {
                it.remove();
            }
        }
    }

    /**
     * Core Consistent Hashing Logic
     * Given a key, find the first node clockwise on the ring.
     */
    public String getNode(String key) {
        if (ring.isEmpty()) {
            return null;
        }

        long hash = hashFunction.hash(key);

        // tailMap(hash) returns all keys >= hash
        // logic: find first key >= hash. If none, wrap around to first key.
        Long targetHash = ring.ceilingKey(hash);
        
        if (targetHash == null) {
            // Wrap around: Key is larger than the largest node hash, so it goes to the first node
            targetHash = ring.firstKey();
        }

        return ring.get(targetHash).getName();
    }
}

// --- 4. Rebalancer & Simulation Module ---
public class ConsistentHashing {

    private static final int NUM_KEYS = 100_000;
    private static final int VNODES = 100;

    public static void main(String[] args) {
        System.out.println("=== Consistent Hashing Simulation (Java) ===\n");

        List<String> keys = new ArrayList<>();
        for (int i = 0; i < NUM_KEYS; i++) {
            keys.add("user_sess_" + i);
        }

        // 1. Initial State
        ConsistentHashRing ring = new ConsistentHashRing(VNODES);
        List<String> initialNodes = Arrays.asList("Node_A", "Node_B", "Node_C");
        
        System.out.println("1. Initializing Ring with 3 Nodes (" + VNODES + " vnodes each)...");
        for (String node : initialNodes) {
            ring.addNode(node);
        }

        // Baseline Mapping
        Map<String, String> mappingBefore = getMapping(ring, keys);
        printDistribution(mappingBefore);

        // 2. Add Node Scenario
        System.out.println("\n2. [SCENARIO] Adding 'Node_D'...");
        ring.addNode("Node_D");
        
        Map<String, String> mappingAfterAdd = getMapping(ring, keys);
        logMovement(mappingBefore, mappingAfterAdd);
        printDistribution(mappingAfterAdd);

        // 3. Remove Node Scenario
        System.out.println("\n3. [SCENARIO] Removing 'Node_A'...");
        ring.removeNode("Node_A");
        
        Map<String, String> mappingAfterRemove = getMapping(ring, keys);
        logMovement(mappingAfterAdd, mappingAfterRemove);
        printDistribution(mappingAfterRemove);
    }

    // --- Helper Methods for Metrics ---

    private static Map<String, String> getMapping(ConsistentHashRing ring, List<String> keys) {
        Map<String, String> mapping = new HashMap<>();
        for (String key : keys) {
            mapping.put(key, ring.getNode(key));
        }
        return mapping;
    }

    private static void printDistribution(Map<String, String> mapping) {
        Map<String, Integer> dist = new HashMap<>();
        for (String node : mapping.values()) {
            dist.put(node, dist.getOrDefault(node, 0) + 1);
        }

        System.out.println("   Load Distribution:");
        
        // Calculate Mean and StdDev
        double mean = (double) mapping.size() / dist.size();
        double sumSquaredDiff = 0.0;
        
        for (Map.Entry<String, Integer> entry : dist.entrySet()) {
            double percent = (entry.getValue() / (double) mapping.size()) * 100;
            System.out.printf("     %s: %d keys (%.2f%%)\n", entry.getKey(), entry.getValue(), percent);
            sumSquaredDiff += Math.pow(entry.getValue() - mean, 2);
        }
        
        double stdDev = Math.sqrt(sumSquaredDiff / dist.size());
        System.out.printf("   Stats: Mean=%.2f, StdDev=%.2f\n", mean, stdDev);
    }

    private static void logMovement(Map<String, String> oldMap, Map<String, String> newMap) {
        int moved = 0;
        Map<String, Integer> flows = new HashMap<>();
        List<String> sampleLogs = new ArrayList<>();

        for (Map.Entry<String, String> entry : oldMap.entrySet()) {
            String key = entry.getKey();
            String oldNode = entry.getValue();
            String newNode = newMap.get(key);

            if (!oldNode.equals(newNode)) {
                moved++;
                String flow = oldNode + " -> " + newNode;
                flows.put(flow, flows.getOrDefault(flow, 0) + 1);
                
                if (sampleLogs.size() < 3) {
                    sampleLogs.add("   [Log] Key '" + key + "' moved: " + flow);
                }
            }
        }

        System.out.println("   Key Movement Analysis:");
        for (String log : sampleLogs) {
            System.out.println(log);
        }
        
        System.out.println("   Movement Summary:");
        for (Map.Entry<String, Integer> entry : flows.entrySet()) {
            System.out.println("     " + entry.getKey() + ": " + entry.getValue() + " keys");
        }

        double percent = (moved / (double) oldMap.size()) * 100;
        System.out.printf("   Total Moved: %d (%.2f%%)\n", moved, percent);
    }
}